### PR TITLE
Bug #369: Add signing and deploying for lsp4j

### PR DESCRIPTION
**Pull request by jonahgraham**
Prior to this the git repo that handled signing was outside
of Eclipse.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -18,16 +18,13 @@ A brief overview of how to use LSP4J to implement a server or a client can be fo
...
-The snapshots are regularly deployed to:
+The snapshots are deployed during the build to:
  * https://oss.sonatype.org/content/repositories/snapshots/org/eclipse/lsp4j/
 
-Also, you can get them directly from the TypeFox Jenkins server:
- * http://services.typefox.io/open-source/jenkins/job/lsp4j/job/master/lastSuccessfulBuild/artifact/build/maven-repository/
-
 ### Supported LSP Versions
 
  * LSP4J 0.7.&ast;, 0.8.&ast; &rarr; LSP 3.14.0
```
Comment (cdietrich) [inline README.md]: list misses 0.9

File: releng/build.Jenkinsfile [hunk 2]
```diff
@@ -0,0 +1,59 @@
+pipeline {
+  agent any
```
Comment (cdietrich) [inline releng/build.Jenkinsfile]: sure this will work on any agent?

File: releng/pom.xml [hunk 3]
```diff
@@ -12,6 +12,12 @@
 		<root-dir>${basedir}/..</root-dir>
 	</properties>
 
+	<pluginRepositories>
```
Comment (cdietrich) [inline releng/pom.xml]: calling `./gradlew generateP2Build` will move these changes to the void again unfortunately

Comment (cdietrich) [reply releng/pom.xml]: https://github.com/TypeFox/gradle-plugins/blob/9742b36882f062d272917528c5f5ccfec3ee21a5/gradle-p2gen/src/main/java/io/typefox/gradle/p2gen/P2GenPlugin.xtend

Comment (cdietrich) [reply releng/pom.xml]: => we would need to fix that there or get rid of the p2gen an maintain the stuff manually as we did in Xtext

File: releng/release-eclipse.sh [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+wget -q --output-document=/dev/null $ARTIFACTS_REPO_TARGET
+
+ECHO=echo
+if [ "$DRY_RUN" == "false" ]; then
+    ECHO=""
+else
+    echo Dry run of build:
+fi
+
+$ECHO mkdir -p $DOWNLOAD_MOUNT
```
Comment (cdietrich) [inline releng/release-eclipse.sh]: this wont do a drops deployment correct?
seems like it was not done for some releases.
am not sure if this is nice for people who want to download the repo
https://download.eclipse.org/lsp4j/downloads/drops/?d

File: gradle/java-compiler-settings.gradle [hunk 5]
```diff
@@ -35,6 +35,49 @@ task javadocJar(type: Jar, dependsOn: javadoc) {
 	from javadoc.destinationDir
 }
 
+ext.signMethod = { jarfile -> 
+	println "Signing $jarfile"
+	def SIGNING_SERVICE = 'http://build.eclipse.org:31338/sign'
+	def STDOUT_FORMAT = '    %{size_upload} bytes uploaded, %{size_download} bytes downloaded (%{time_total} s)\\n'
+
+	ProcessBuilder curl_pb = new ProcessBuilder("curl", 
+		"--fail", "--silent", "--show-error", "--output",
```
Comment (cdietrich) [inline gradle/java-compiler-settings.gradle]: we had to add some retrys to Xtext here as the service is instable sometimes

File: releng/pom.xml (hunk 3 above)
Comment (jonahgraham) [reply releng/pom.xml]: OK. I hadn't realized it was being generated (TODO add "this file is auto-generated" comment)

Comment (jonahgraham) [reply releng/pom.xml]: For xtext did you just start with the generated file and remove the generateP2Build?

File: gradle/java-compiler-settings.gradle (hunk 5 above)
Comment (cdietrich) [reply gradle/java-compiler-settings.gradle]: https://git.eclipse.org/c/cbi/org.eclipse.cbi.git/tree/maven-plugins/eclipse-jarsigner-plugin/src/main/java/org/eclipse/cbi/maven/plugins/jarsigner/mojo/SignMojo.java does 3 retrys too

File: releng/release-eclipse.sh (hunk 4 above)
Comment (jonahgraham) [reply releng/release-eclipse.sh]: The "drops" is a separate issue and if someone needs it they can request and I will look into doing that. Do you need it @cdietrich?

File: releng/pom.xml (hunk 3 above)
Comment (cdietrich) [reply releng/pom.xml]: yes.

Comment (cdietrich) [reply releng/pom.xml]: i am not sure if we can change the typefox plugin to add the repo and profile

File: gradle/java-compiler-settings.gradle (hunk 5 above)
Comment (jonahgraham) [reply gradle/java-compiler-settings.gradle]: Thanks, I had seen some retries in other code - I can look at that once we have some instability - at the moment the full build time is very short so running it again is less time consuming. I assume xtext has a much longer build time, so having a signing failure would be much more inconvenient?

File: releng/release-eclipse.sh (hunk 4 above)
Comment (cdietrich) [reply releng/release-eclipse.sh]: not right now

File: releng/build.Jenkinsfile (hunk 2 above)
Comment (jonahgraham) [reply releng/build.Jenkinsfile]: Any agent that lsp4j CI currently has. If (when!) it is moved to JIRO more will be needed here.

File: README.md (hunk 1 above)
Comment (jonahgraham) [reply README.md]: Fixed.

File: gradle/java-compiler-settings.gradle (hunk 5 above)
Comment (cdietrich) [reply gradle/java-compiler-settings.gradle]: indeed

**Comment by cdietrich** [comment]
@jonahgraham will this do deployments for PRs too or for branches only?

File: releng/pom.xml (hunk 3 above)
Comment (jonahgraham) [reply releng/pom.xml]: I have changed this to be manually generated and removed generateP2Build.

Comment (cdietrich) [reply releng/pom.xml]: ok. can you please  create a follow up task to discuss this with @spoenemann or @dhuebner  once they are back?

**Comment by jonahgraham** [comment]
> @jonahgraham will this do deployments for PRs too or for branches only?

It will do deploy to download.eclipse.org for branches in this repo - but it will not deploy to maven central except for master and release_* named branches. See https://github.com/eclipse/lsp4j/pull/378/files#diff-c89773d0f8e78cce85e72b37c626d0a7R23

File: gradle/p2-deployment.gradle [hunk 6]
```diff
@@ -1,53 +0,0 @@
...
- * This program and the accompanying materials are made available under the
- * terms of the Eclipse Public License v. 2.0 which is available at
- * http://www.eclipse.org/legal/epl-2.0,
- * or the Eclipse Distribution License v. 1.0 which is available at
- * http://www.eclipse.org/org/documents/edl-v10.php.
- * 
- * SPDX-License-Identifier: EPL-2.0 OR BSD-3-Clause
- ******************************************************************************/
-
-apply plugin: 'io.typefox.gradle.p2gen'
```
Comment (cdietrich) [inline gradle/p2-deployment.gradle]: plugin is still loaded in build.gradle

**Comment by cdietrich** [comment]
ok. we might need a follow up task for removal of branches later

File: gradle/p2-deployment.gradle (hunk 6 above)
Comment (jonahgraham) [reply gradle/p2-deployment.gradle]: I think I removed it: https://github.com/eclipse/lsp4j/pull/378/files#diff-c197962302397baf3a4cc36463dce5eaL27

Comment (cdietrich) [reply gradle/p2-deployment.gradle]: i mean this one https://github.com/eclipse/lsp4j/blob/ec16cc8b2dfa797085be596fc6dd743ea5a4bc65/build.gradle#L20

Comment (jonahgraham) [reply gradle/p2-deployment.gradle]: Removed too: https://github.com/eclipse/lsp4j/pull/378/files#diff-c197962302397baf3a4cc36463dce5eaL20

Comment (jonahgraham) [reply gradle/p2-deployment.gradle]: PS I don't know how you made your link to the source code look pretty (with inlined source) and mine are ugly links.

**Comment by cdietrich** [comment]
ok. in general this looks ok to me.
i still have no clue who maintains
e.g.
http://download.eclipse.org/lsp4j/updates/releases/compositeContent.jar
http://download.eclipse.org/lsp4j/updates/releases/compositeArtifacts.jar

File: gradle/p2-deployment.gradle (hunk 6 above)
Comment (cdietrich) [reply gradle/p2-deployment.gradle]: i used the rightklick copy permalink

**Comment by jonahgraham** [comment]
> i still have no clue who maintains

Personally I would propose removing them. I can't understand in what circumstances a consumer would correctly want that composite. There is an argument for having composites that cover 0.8.0 and 0.8.1 in the same composite though, but that is not what is in there.

**Comment by cdietrich** [comment]
the only usecase is basically: always take the latest and hope the best.
in Xtext we have a composite forever. and we have adapted the jobs to adapt them
but we dont have jars but xmls.
=> maybe we should leave it for now and create another follow up tasks
when we know how it was created in the past we might be able to replicate
it.

File: gradle/p2-deployment.gradle (hunk 6 above)
Comment (jonahgraham) [reply gradle/p2-deployment.gradle]: Thanks - that isn't an option in the diff viewer, but for completeness on this resolved conversation here is the link in the new file showing it gone:

https://github.com/eclipse/lsp4j/blob/fe5212dd8a2a7bed66cc30c669a9aac58226ef34/build.gradle#L18-L20

**Pull request merged**